src/v2/params/types.ts

```diff
@@ -69,31 +72,36 @@ export interface SelectOptions<T = unknown> {
   value: T;
 }
 
-export interface ParamSpec<T = unknown> {
+export type ParamSpec<T = unknown> = {
   name: string;
   default?: T;
   label?: string;
-  as?: string;
   description?: string;
   type: ParamValueType;
-  input?: ParamInput<T>;
-}
+} & HasAtMostOneInput<T>;
 
 export type ParamOptions<T = unknown> = Omit<ParamSpec<T>, 'name' | 'type'>;
 
-export class Param<T = unknown> {
+export abstract class Param<T extends string | number | boolean | string[]> {
```

Comment:
I want to play with abstract classes a bit to see if it changes my mind about the class hierarchy. I guess I don't mind having all expression types supporting the `eq` method even though I loose my sanity a smidge when I see "if (bool === true)". But Boolean expressions need to be their own type I think because only boolean expressions can implement `then`.  So maybe.... create a boolean expression type that implements `then` and make sure `eq` returns it and that `eq` is on BooleanParam?